### PR TITLE
Add data table

### PR DESCRIPTION
**Pull request by balakrishnan-s5**
**Changes**

- Added component data table

   [Delete] branch

**Additional information**

- You may refer to this [BOOKINGSG-4436](https://jira.ship.gov.sg/browse/BOOKINGSG-4436) 
- [Figma link](https://www.figma.com/file/us90jOpWBahsqK2VAluPvD/Flagship-Design-System?type=design&node-id=10746-81730&mode=design&t=okPG2EZGHJJ8be4q-0)

### CODE REVIEW

File: src/data-table/types.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    rowsConfig?: RowsConfigProps | undefined;
+    rows?: RowProps[] | undefined;
+    className?: string | undefined; // css class to put on table
+    selectionConfig?: SelectionConfigProps | undefined;
+    selection?: string[] | undefined; // ids of all selected rows
+    sortIndicators?: SortIndicatorsProps | undefined; // columns that want to show a sort indicator
+    actionsConfig?: ActionsConfigProps | undefined;
+    alternatingRows?: boolean | undefined;
+    customEmptyView?: () => ReactNode | string | undefined;
+    isLoadingData: boolean | undefined;
```
Comment (qroll) [inline src/data-table/types.ts:16]: some flexibility for other potential statuses like error in the future
```suggestion
    loadState?: "loading" | "success" | undefined;
```

File: src/data-table/types.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import { CSSProperties, ReactNode } from "react";
+
+export interface DataTableProps {
+    testid?: string | undefined;
+    headersConfig?: HeadersConfigProps | undefined;
+    headers: HeadersProps[];
+    rowsConfig?: RowsConfigProps | undefined;
+    rows?: RowProps[] | undefined;
+    className?: string | undefined; // css class to put on table
```
Comment (qroll) [inline src/data-table/types.ts:9]: set these comments as jsdocs instead
```suggestion
    /** css class to put on table **/
    className?: string | undefined;
```

File: src/data-table/types.ts
Comment (qroll) [inline src/data-table/types.ts:1]: more of a DX question, can take this back for discussion with RBS and your team since you guys will be consuming this component

would devs prefer to work with separate props to configure header and row (current state)

or have a combined config e.g.

```
columns={[
      {
        header: "Status", // or ReactNode
        fieldKey: "status", // maps to the object property
        sort: "asc", // or "desc"
        onHeaderClick: () => {},
        renderRowCell: (value, item, renderProps) => {},
      }
]}
```

**Review by qroll — COMMENTED** [review]
some initial comments

File: src/data-table/types.ts
Comment (balakrishnan-s5) [reply src/data-table/types.ts:1]: As discussed we will keep the header and row configs seprated

Comment (keithtxw) [reply src/data-table/types.ts:1]: Still find the current approach awkward. Puts alot of onus on the user to make sure they specify the matching number of header cols and row cols. Let me think more if there is a better way

File: src/data-table/data-table.styles.tsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    width: 100%;
+    border: 2px solid ${Color.Neutral[6]};
+    border-radius: 0.5rem 0.5rem 0 0;
+    overflow: auto;
+`;
+
+export const Table = styled.table<{
+    $addMarginToFirstColumn?: boolean;
+    $addMarginToLastColumn?: boolean;
+}>`
```
Comment (qroll) [inline src/data-table/data-table.styles.tsx:20]: for these typings, pull them out into interfaces e.g. `TableStyleProps`

do add the STYLE INTERFACE header as well

File: src/data-table/data-table.styles.tsx
Comment (qroll) [inline src/data-table/data-table.styles.tsx:1]: is it intended for the non-first cells to not have horizontal padding? they are misaligned with the headers which do have padding

https://github.com/LifeSG/react-design-system/assets/16853062/c9864510-0ba5-41bb-ac37-ed529c1cfa0f

File: src/data-table/data-table.styles.tsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    display: flex;
+    align-items: center;
+    justify-content: center;
+`;
+
+export const CheckBox = styled(Checkbox)``;
+
+const sortIconStyles = css`
+    color: ${Color.Neutral[1]};
+    font-weight: bold;
```
Comment (qroll) [inline src/data-table/data-table.styles.tsx:119]: this can be removed as there's no font weight for icons

File: src/data-table/data-table.styles.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+export const EmptyDataElement = styled.div`
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    gap: 2rem;
+    opacity: 0;
+    img {
+        width: 400px;
+        height: 280px;
```
Comment (qroll) [inline src/data-table/data-table.styles.tsx:145]: convert to `rem` units

File: src/data-table/data-table.styles.tsx
Comment (qroll) [reply src/data-table/data-table.styles.tsx:1]: similarly the loader and empty image are not centered

Screenshot 2023-08-22 at 9 38 34 AM

Comment (qroll) [reply src/data-table/data-table.styles.tsx:1]: the last column should have the additional right padding as well to match the first column

Screenshot 2023-08-22 at 9 39 23 AM

File: src/data-table/data-table.styles.tsx [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        return Color.Neutral[7];
+    } else {
+        return Color.Neutral[8];
+    }
+};
+
+export const BodyRow = styled.tr<{
+    alternating: boolean;
+    isSelected?: boolean;
+    isSelectable?: boolean;
```
Comment (qroll) [inline src/data-table/data-table.styles.tsx:84]: to add `$` prefix as these are not native html attributes. do update other instances of this

File: src/data-table/types.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    /** columns that want to show a sort indicator **/
+    sortIndicators?: SortIndicatorsProps | undefined;
+    // actionsConfig?: ActionsConfigProps | undefined;
+    alternatingRows?: boolean | undefined;
+    customEmptyView?: () => ReactNode | string | undefined;
+    loadState: LoadingType | undefined;
+    emptyView?: ErrorDisplayAttributes | undefined;
+}
+
+export type LoadingType = "loading" | "success" | "error";
```
Comment (qroll) [inline src/data-table/types.ts:24]: if there's no error state specification, we can leave it out for now to avoid unintended usage

File: src/data-table/data-table.tsx [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+                        }}
+                    />
+                </CheckBoxWrapper>
+            </BodyCell>
+        );
+    };
+
+    const basicEmptyView = () => {
+        return (
+            <ErrorDisplay
```
Comment (qroll) [inline src/data-table/data-table.tsx:249]: is there a default title/description?

also note that the ErrorDisplay currently doesn't allow empty title/description (it will revert to the placeholder copy). if this functionality is needed, please make the change in a separate PR

separately, the header in ErrorDisplay is H1 but Figma shows H3 for the table. can check with UX if this is ok, otherwise you can override the text styling

**Review by qroll — CHANGES_REQUESTED** [review]

**Review by keithtxw — CHANGES_REQUESTED** [review]

File: src/data-table/data-table.styles.tsx [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+`;
+
+export const HeaderCell = styled.th<HeaderCellProps>`
+    padding: 1rem 0;
+    text-align: left;
+    cursor: ${(props) => (props.$clickable ? "pointer" : "default")};
+    max-width: ${(props) => (props.$maxWidth ? props.$maxWidth : "auto")};
+    vertical-align: middle;
+    ${TextStyleHelper.getFontFamily("H5", "bold")}
+    color: ${Color.Neutral[1]}
```
Comment (qroll) [inline src/data-table/data-table.styles.tsx:85]: to add the semicolon here

File: src/data-table/data-table.styles.tsx [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    vertical-align: middle;
+`;
+
+export const CheckBoxWrapper = styled.div`
+    display: flex;
+    align-items: center;
+    justify-content: center;
+`;
+
+export const CheckBox = styled(Checkbox)``;
```
Comment (qroll) [inline src/data-table/data-table.styles.tsx:142]: can be removed if no special styling is required

File: src/data-table/data-table.styles.tsx
Comment (qroll) [reply src/data-table/data-table.styles.tsx:1]: - [ ] the last column should have a right padding

Screenshot 2023-08-28 at 11 41 34 AM

File: src/data-table/data-table.styles.tsx [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+export const LoaderWrapper = styled.div`
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+`;
+
+export const EmptyDataElement = styled.div`
```
Comment (qroll) [inline src/data-table/data-table.styles.tsx:150]: this seems to be unused

**Review by qroll — COMMENTED** [review]
noting down the outstanding items that should be covered separately:
- props changes and error display enhancements (follow up PR)
- style changes to match Figma (part 2 PR)
- any new features (part 3 PR)
- storybook documentation (final PR if needed)

File: src/data-table/data-table.styles.tsx [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+        }
+    }
+`;
+
+export const CustomErrorDisplay = styled(ErrorDisplay)`
+    div {
+        h1 {
+            ${TextStyleHelper.getFontFamily("H3", "bold")}
+        }
+    }
```
Comment (qroll) [inline src/data-table/data-table.styles.tsx:176]: not ideal way to style the text elements. let's clarify with UX if h3 is the standard size for an inline error display. we can introduce a prop like `displayType: "default" | "small"` as part of the follow up PR.

Comment (SasithaRajapakasha) [reply src/data-table/data-table.styles.tsx:176]: I'll remove the font size and weight change. Follow up URL, I'll get update from UX and do the changes.

**Review by qroll — APPROVED** [review]
